File: src/xyz/jadonfowler/sbhs/SpriteManager.java
package xyz.jadonfowler.sbhs;

import java.awt.*;
import java.awt.event.*;
import java.awt.image.*;
import java.io.*;
import java.util.*;
import javax.imageio.*;
import javax.swing.*;

public class SpriteManager {
    public static HashMap<String, HashMap<String, BufferedImage>> SPRITES = new HashMap<String, HashMap<String, BufferedImage>>();

    public static void addCharacterSpriteTab(JTabbedPane pane, String name, int offset) throws Exception {
        //@formatter:off
        SPRITES.put(name, new HashMap<String, BufferedImage>());
        JTabbedPane t = new JTabbedPane();
        // When I try to condense the equations into numbers, I get weird
        // errors. Probably has to do with the fact that SEGA is so idiotic that
        // they put giant spaces in between certain animations for no reason
        // whatsoever. It was nice for the first few animations, but then I
        // noticed that not all of them had it, which makes me confused as
        // to why they were there to begin with. Thank you SEGA, for making
        // a game more confusing that Minecraft's protocol.
        t.addTab("Idle", null, createSpritePanel(name, "Idle", //6
                offset, 32 * 7 - 8), 
                "Edit " + name + " Idle Sprite");
        // Why the hell there is a "jog" animation?! JUST GO INTO RUNNING
        // ALREADY SEGA, YOU SURE HAVEN'T DONE THAT IN A WHILE.
        t.addTab("Jog", null, createSpritePanel(name, "Jog", //1
                offset + (64 * 2 * 64) + (64 * 16), 32),
                "Edit " + name + " Jog Sprite");
        t.addTab("Run", null, createSpritePanel(name, "Run", //8
                offset + (64 * 3 * 64) + (64 * 24), 32 * 9),
                "Edit " + name + " Run Sprite");
        t.addTab("Halt", null, createSpritePanel(name, "Halt", //4
                offset + (64 * 5 * 64) + (64 * 40), 32 * 4 + 16),
                "Edit " + name + " Halt Sprite");
        t.addTab("Dash", null, createSpritePanel(name, "Dash", //7
                offset + (64 * 6 * 64) + (64 * 48), 32 * 8),
                "Edit " + name + " Dash Sprite");
        t.addTab("Turn", null, createSpritePanel(name, "Turn", //3
                offset + (64 * 9 * 64), 32 * 4 - 16),
                "Edit " + name + " Turn Sprite");
        t.addTab("Change Direction", null, createSpritePanel(name, "Change Direction", //4
                offset + (64 * 10 * 64) + (64 * 8), 32 * 5 - 16),
                "Edit " + name + " Change Direction Sprite");
        t.addTab("Fall", null, createSpritePanel(name, "Fall", //4
                offset + (64 * 11 * 64) + (64 * 16), 32 * 5 - 16),
                "Edit " + name + " Fall Sprite");
        t.addTab("Jump", null, createSpritePanel(name, "Jump", //5
                offset + (64 * 12 * 64) + (64 * 24), 32 * 6 - 16),
                "Edit " + name + " Jump Sprite");
        t.addTab("Land", null, createSpritePanel(name, "Land",
                offset + (64 * 14 * 64) + (64 * 38), 32 * 4 - 16), //3
                "Edit " + name + " Land Sprite");
        t.addTab("Double Jump", null, createSpritePanel(name, "Double Jump",
                offset + (64 * 15 * 64) + (64 * 48), 32 * 8), //7
                "Edit " + name + " Double Jump");
        // These are the normal `B` attacks
        t.addTab("Attack", null, createSpritePanel(name, "Attack 1",
                offset + (64 * 18 * 64), 32 * 7 - 8), //6
                "Edit " + name + " Attack 1");
        t.addTab("Attack 2", null, createSpritePanel(name, "Attack 2",
                offset + (64 * 20 * 64) + (64 * 16), 32 * 7 - 8), //6
                "Edit " + name + " Attack 2");
        t.addTab("Attack 3", null, createSpritePanel(name, "Attack 3",
                offset + (64 * 22 * 64) + (64 * 32), 32 * 8 - 4), //7
                "Edit " + name + " Attack 3");
        t.addTab("Big Attack", null, createSpritePanel(name, "Big Attack",
                /* This one is a little weird */
                offset + (64 * 24 * 64) + (64 * 28), 32 * 19 + 8), //17
                "Edit " + name + " Big Attack");
        // When you press the direction away from where you're pointing and B
        t.addTab("Back Attack", null, createSpritePanel(name, "Back Attack",
                offset + (64 * 29 * 64) + (64 * 16), 32 * 10), //9
                "Edit " + name + " Back Attack");
        pane.addTab(name, null, t, "Edit " + name + " Sprite");
        //@formatter:on
    }

    public static void addSpriteTab(JTabbedPane pane, String name, int offset, int amount) throws Exception {
        SPRITES.put(name, new HashMap<String, BufferedImage>());
        pane.addTab(name, null, createSpritePanel(name, "Idle", offset, amount), "Edit " + name + " Sprite");
    }

    public static JPanel createSpritePanel(String name, String state, int offset, int amount) throws Exception {
        BufferedImage img = new BufferedImage(8 * 4, 8 * (int) Math.ceil(amount / 4), BufferedImage.TYPE_INT_RGB);
        Graphics2D g = img.createGraphics();
        g.setColor(Color.RED);
        g.fillRect(0, 0, img.getWidth(), img.getHeight());
        g.dispose();
        int x = 1, y = 1, s = 0;
        for (int i = offset, size = 4, rows = 8; i < offset + (size * rows * amount); i++) {
            SBHS.raf.seek(i);
            int v = SBHS.raf.read();
            String value = SBHS.reverse((v < 0x10 ? "0" : "") + Integer.toString(v, 16));
            Color o1 = Color.black, o2 = Color.white;
            //System.out.print(value + " : ");
            int v1 = Integer.parseInt(value.toCharArray()[0] + "", 16);
            int v2 = Integer.parseInt(value.toCharArray()[1] + "", 16);
            //System.out.println(v1 + " " + v2);
            try {
                //printPalette(PaletteManager.PALETTES.get(name));
                String s1 = PaletteManager.PALETTES.get(name)[v1];
                if (s1 == null) s1 = "0000";
                o1 = GBAColor.fromGBA(s1);
                String s2 = PaletteManager.PALETTES.get(name)[v2];
                if (s2 == null) s2 = "0000";
                o2 = GBAColor.fromGBA(s2);
            }
            catch (Exception e) {
                e.printStackTrace();
                System.exit(-1);
            }
            int col1 = o1.getRGB();
            int col2 = o2.getRGB();
            // if (v1 == 0) System.out.println(o1);
            // System.out.println(
            // " S: " + s + " X: " + x + " Y: " + y + " MaxX: " + img.getWidth()
            // + " MaxY: " + img.getHeight());
            try {
                img.setRGB(x - 1, y - 1, col1);
            }
            catch (Exception e) {
                System.out.println("ERROR @ " + x + " " + y + " / " + img.getWidth() + " " + img.getHeight());
                e.printStackTrace();
                System.exit(-1);
            }
            x++;
            // System.out.println(
            // " S: " + s + " X: " + x + " Y: " + y + " MaxX: " + img.getWidth()
            // + " MaxY: " + img.getHeight());
            try {
                img.setRGB(x - 1, y - 1, col2);
            }
            catch (Exception e) {
                System.out.println("ERROR @ " + x + " " + y + " / " + img.getWidth() + " " + img.getHeight());
                e.printStackTrace();
                System.exit(-1);
            }
            if (x != 0 && x % 8 == 0) {
                x -= 8;
                if (y != 0 && y % 8 == 0) {
                    y -= 8;
                    x += 8;
                    s++;
                    if (s == 4) {
                        s = 0;
                        x = 0;
                        y += 8;
                    }
                }
                y++;
            }
            x++;
        }
        SPRITES.get(name).put(state, img);
        JButton write = new JButton("Write to ROM");
        write.addActionListener(new ActionListener() {
            @Override public void actionPerformed(ActionEvent e) {
                writeImage(name, state, offset, amount);
            }
        });
        JButton save = new JButton("Save sprites");
        save.addActionListener(new ActionListener() {
            @Override public void actionPerformed(ActionEvent e) {
                System.out.println("Saving image " + name + ": " + state);
                BufferedImage i = SPRITES.get(name).get(state);
                JFileChooser fc = new JFileChooser();
                if (fc.showSaveDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                    try {
                        File o = fc.getSelectedFile();
                        ImageIO.write(i, "png", o);
                    }
                    catch (IOException x) {
                        x.printStackTrace();
                    }
                }
            }
        });
        JButton upload = new JButton("Upload sprites");
        upload.addActionListener(new ActionListener() {
            @Override public void actionPerformed(ActionEvent e) {
                System.out.println("Uploading image " + name + ": " + state);
                JFileChooser fc = new JFileChooser();
                if (fc.showOpenDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                    try {
                        BufferedImage i = ImageIO.read(fc.getSelectedFile());
                        SPRITES.get(name).put(state, i);
                    }
                    catch (IOException x) {
                        x.printStackTrace();
                    }
                }
            }
        });
        JPanel jp = new JPanel();
        int h = (int) Math.ceil(amount / 4);
        int sw = 200 / h * 16;
        int sh = SBHS.frame.getHeight() - 64;
        ImageIcon spriteSheet = new ImageIcon(scale(img, sw, sh));
        jp.add(write);
        jp.add(save);
        jp.add(upload);
        jp.add(new JLabel(spriteSheet));
        return jp;
    }
    
    /**
     * Writes uncompressed sprite to ROM
     * @param img Image to write
     * @param offset Offset of image in ROM
     * @param amount Amount of 8x8 squares
     */
    public static void writeImage(String name, String state, int offset, int amount) {
        BufferedImage img = SPRITES.get(name).get(state);
        printPalette(PaletteManager.PALETTES.get(name));
        int x = 1, y = 1, s = 0;
        for (int i = offset, size = 4, rows = 8; i < offset + (size * rows * amount); i++) {
            Color c1 = getColorFromImage(img, x - 1, y - 1);
            // This gets the index of the color in the palette,
            // which is the value we need to write to the ROM
            //System.out.println("Palette contains " + c1 + ": " + Arrays.asList(PaletteManager.PALETTES.get(name)).contains(c1));
            int v1 = getIndexFromPalette(name, c1);
            x++;
            Color c2 = getColorFromImage(img, x - 1, y - 1);
            if (x != 0 && x % 8 == 0) {
                x -= 8;
                if (y != 0 && y % 8 == 0) {
                    y -= 8;
                    x += 8;
                    s++;
                    if (s == 4) {
                        s = 0;
                        x = 0;
                        y += 8;
                    }
                }
                y++;
            }
            x++;
            //System.out.println("Palette contains " + c2 + ": " + Arrays.asList(PaletteManager.PALETTES.get(name)).contains(c2));
            int v2 = getIndexFromPalette(name, c2);
            // Writing to ROM
//            if (v1 < 0) v1 = 0;
//            if (v2 < 0) v2 = 0;
            int v = Integer.parseInt(Integer.toString(v2, 16) + "" + Integer.toString(v1, 16), 16);
            try {
                SBHS.raf.seek(i);
                SBHS.raf.write(v);
                System.out.println("Colors: " + c1 + " " + c2 + " Wrote: " + Integer.toString(v, 16) + " @ 0x"
                        + Integer.toString(i, 16));
            }
            catch (IOException e) {
                e.printStackTrace();
            }
        }
        printPalette(PaletteManager.PALETTES.get(name));
    }

    /**
     * Scales image to new size
     * @param src Image to scale
     * @param w Width of new image
     * @param h Height of new image
     * @return Scaled image
     */
    public static BufferedImage scale(BufferedImage src, int w, int h) {
        BufferedImage img = new BufferedImage(w, h, BufferedImage.TYPE_INT_RGB);
        int x, y;
        int ww = src.getWidth();
        int hh = src.getHeight();
        for (x = 0; x < w; x++) {
            for (y = 0; y < h; y++) {
                int col = src.getRGB(x * ww / w, y * hh / h);
                img.setRGB(x, y, col);
            }
        }
        return img;
    }
    
    /**
     * Convert RGB from image pixel to GBA Color
     * @param img Source image
     * @param x X of pixel
     * @param y Y of pixel
     * @return GBA Color of pixel's color
     */
    public static Color getColorFromImage(BufferedImage img, int x, int y) {
        int rgb = img.getRGB(x, y);
        Color c = new Color(rgb);
        System.out.print(" " + c + "(" + GBAColor.toGBA(c) + ") ");
        return c;
    }

    protected static void printPalette(String[] s) {
        System.out.print("[");
        for (int i = 0; i < s.length; i++) {
            System.out.print(s[i] + "(");
            System.out.print(GBAColor.fromGBA(s[i]) + ")");
            if (i < s.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }

    public static int getIndexFromPalette(String name, Color color) {
        ArrayList<String> p = new ArrayList<String>(Arrays.asList(PaletteManager.PALETTES.get(name)));
        for (String s : p) {
            int i = p.indexOf(s);
            Color pc = GBAColor.fromGBA(s);
            if (pc.equals(color)) return i;
        }
        System.out.print("Error finding color " + color);
        return 0;
    }
}

File: src/main/kotlin/xyz/jadonfowler/sbhs/SpriteManager.kt
package xyz.jadonfowler.sbhs

import java.awt.Color
import java.awt.ScrollPane
import java.awt.image.BufferedImage
import java.io.IOException
import java.util.*
import javax.imageio.ImageIO
import javax.swing.*
import javax.swing.BoxLayout


class Tuple<out A>(val a: A, val b: A) {
    operator fun get(i: Int): A {
        return when (i) {
            0 -> a
            1 -> b
            else -> throw IllegalArgumentException("$i is not in this Tuple.")
        }
    }
}

object SpriteManager {

    class SpriteSection(val values: Array<Array<Int>>)

    var SPRITES = HashMap<String, BufferedImage>()

    @Throws(Exception::class)
    fun addCharacterSpriteTab(pane: JTabbedPane, name: String, spriteOffset: Int, paletteOffset: Int) {
        val spriteData = listOf(
                // One frame is 0x480
                Tuple(spriteOffset, 8), // Idle 6
                Tuple(spriteOffset + 0x02400, 1), // Jog 1
                Tuple(spriteOffset + 0x03600, 8), // Run 8
                Tuple(spriteOffset + 0x05a00, 4), // Halt 4
                Tuple(spriteOffset + 0x06c00, 8), // Dash 7
                Tuple(spriteOffset + 0x09000, 4), // Turn 3
                Tuple(spriteOffset + 0x0a200, 4), // Change Direction 4
                Tuple(spriteOffset + 0x0b400, 4), // Fall 4
                Tuple(spriteOffset + 0x0c600, 8), // Jump 5
                Tuple(spriteOffset + 0x0ea00, 4), // Land 3
                Tuple(spriteOffset + 0x0fc00, 8), // Double Jump 7
                Tuple(spriteOffset + 0x12000, 8), // Attack 1 6
                Tuple(spriteOffset + 0x14400, 8), // Attack 2 6
                Tuple(spriteOffset + 0x16800, 8), // Attack 3 7
                Tuple(spriteOffset + 0x18c00, 16), // Big Attack 17
                Tuple(spriteOffset + 0x1d400, 12), // Back Attack 9
                Tuple(spriteOffset + 0x20a00, 12),
                Tuple(spriteOffset + 0x24000, 8),
                Tuple(spriteOffset + 0x26400, 12),
                Tuple(spriteOffset + 0x29a00, 8),
                Tuple(spriteOffset + 0x2be00, 8),
                Tuple(spriteOffset + 0x2e200, 16),
                Tuple(spriteOffset + 0x32a00, 8),
                Tuple(spriteOffset + 0x34e00, 12),
                Tuple(spriteOffset + 0x38400, 8),
                Tuple(spriteOffset + 0x3a800, 8),
                Tuple(spriteOffset + 0x3cc00, 4),
                Tuple(spriteOffset + 0x3de00, 8),
                Tuple(spriteOffset + 0x40200, 4),
                Tuple(spriteOffset + 0x41400, 1),
                Tuple(spriteOffset + 0x42600, 8),
                Tuple(spriteOffset + 0x44a00, 8),
                Tuple(spriteOffset + 0x46e00, 12),
                Tuple(spriteOffset + 0x4a400, 4),
                Tuple(spriteOffset + 0x4b600, 8),
                Tuple(spriteOffset + 0x4da00, 1),
                Tuple(spriteOffset + 0x4ec00, 4)
        )
        pane.addTab(name, null, createSpritePanel(name, spriteData, paletteOffset), "Edit $name Sprite")
    }

    @Throws(Exception::class)
    fun addSpriteTab(pane: JTabbedPane, name: String, offset: Int, amount: Int) {
        pane.addTab(name, null, createSpritePanel(name, listOf(Tuple(offset, amount)), -1, 4), "Edit $name Sprite")
    }

    @Throws(Exception::class)
    fun createSpritePanel(name: String, spriteData: List<Tuple<Int>>, paletteOffset: Int, size: Int = 6): JPanel {
        val maxFrames = spriteData.map { it[1] }.max() ?: 6 /*default frame count, though this should never be null*/

        val imgWidth = 8 * size * spriteData.size
        val imgHeight = 8 * size * maxFrames
        val img = BufferedImage(imgWidth, imgHeight, BufferedImage.TYPE_INT_RGB)

        val g = img.createGraphics()
        g.color = GBAColor.fromGBA(PaletteManager.PALETTES[name]!![0])
        g.fillRect(0, 0, img.width, img.height)
        g.dispose()

        spriteData.forEachIndexed { animationIndex, animationData ->
            val offset = animationData[0]
            var currentFrame = 0
            while (currentFrame < animationData[1]) {
                val sections = Array(36, { SpriteSection(Array(8, { Array(8, { -1 }) })) })

                var currentSection = 0
                var x = 1
                var y = 1
                val frameOffset = offset + size * size * currentFrame * 32
                var i = frameOffset
                while (i < frameOffset + 36 * 32) {
                    // Getting the values
                    SBHS.raf.seek(i.toLong())
                    val v = SBHS.raf.read()
                    val value = ((if (v < 0x10) "0" else "") + Integer.toString(v, 16)).reversed()
                    val v1 = Integer.parseInt(value[0].toString(), 16)
                    val v2 = Integer.parseInt(value[1].toString(), 16)

                    // Setting the values
                    sections[currentSection].values[y - 1][x - 1] = v1
                    x++
                    sections[currentSection].values[y - 1][x - 1] = v2

                    // Check bounds
                    if (x != 0 && x % 8 == 0) {
                        x -= 8
                        if (y != 0 && y % 8 == 0) {
                            x = 0
                            y = 0
                            currentSection++
                        }
                        y++
                    }
                    x++
                    i++
                }

                /**
                0123OP
                4567QR
                89ABST
                CDEFUV
                GHIJWX
                KLMNYZ
                 */
                val sortedSections = arrayOf(
                        sections[0], sections[1], sections[2], sections[3], sections[24], sections[25],
                        sections[4], sections[5], sections[6], sections[7], sections[26], sections[27],
                        sections[8], sections[9], sections[10], sections[11], sections[28], sections[29],
                        sections[12], sections[13], sections[14], sections[15], sections[30], sections[31],
                        sections[16], sections[17], sections[18], sections[19], sections[32], sections[33],
                        sections[20], sections[21], sections[22], sections[23], sections[34], sections[35]
                )

                var sy = 0
                sortedSections.forEachIndexed { i, spriteSection ->
                    spriteSection.values.forEachIndexed { y, values ->
                        values.forEachIndexed { x, v ->
                            var s: String? = PaletteManager.PALETTES[name]!![v]
                            if (s == null) s = "0000"
                            val c = GBAColor.fromGBA(s)
                            val ix = x + (i % size) * 8 + (8 * size * animationIndex)
                            val iy = y + sy * 8 + size * (size + 2) * currentFrame
                            try {
                                img.setRGB(ix, iy, c.rgb)
                            } catch(e: Exception) {
//                                println("error @ ($ix, $iy) / (${img.width}, ${img.height})")
//                                e.printStackTrace()
//                                System.exit(1)
                            }
                        }

                    }
                    if (i != 0 && (i + 1) % size == 0) sy++
                }
                currentFrame++
            }
        }

        // Write palette to upper left corner
        PaletteManager.PALETTES[name]!!.forEachIndexed { i, s ->
            val color = GBAColor.fromGBA(s)
            img.setRGB(i, 0, color.rgb)
        }

        SPRITES[name] = img
        val write = JButton("Write to ROM")
        write.addActionListener { writeImage(name, spriteData, paletteOffset); println("Done writing to $name.") }
        val save = JButton("Save sprites")
        save.addActionListener {
            println("Saving Image: $name")
            val i = SPRITES[name]!!
            val fc = JFileChooser()
            if (fc.showSaveDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                try {
                    val o = fc.selectedFile
                    ImageIO.write(i, "png", o)
                } catch (x: IOException) {
                    x.printStackTrace()
                }

            }
        }
        val upload = JButton("Upload sprites")
        upload.addActionListener {
            println("Uploading Image: $name")
            val fc = JFileChooser()
            if (fc.showOpenDialog(SBHS.frame) == JFileChooser.APPROVE_OPTION) {
                try {
                    val i = ImageIO.read(fc.selectedFile)
                    SPRITES[name] = i
                } catch (x: IOException) {
                    x.printStackTrace()
                }

            }
        }
        val jp = JPanel()
        val sw = ((SBHS.frame.width - 64) * 0.7).toInt()
        val sh = (sw * img.height) / img.width
//        val spriteSheet = ImageIcon(scale(img, sw, sh))
        val spriteSheet = ImageIcon(img)

        val buttons = JPanel()
        buttons.layout = BoxLayout(buttons, BoxLayout.Y_AXIS)
        buttons.add(write)
        buttons.add(save)
        buttons.add(upload)
        jp.add(buttons)

        val spriteLabel = JLabel(spriteSheet)
        val scrollPane = ScrollPane()
        scrollPane.setBounds(0, 0, (SBHS.frame.width * 0.9).toInt(), (SBHS.frame.height * 0.9).toInt())
        scrollPane.add(spriteLabel)
        jp.add(scrollPane)
        return jp
    }

    /**
     * Writes uncompressed sprite to ROM
     *
     * @param spriteData List of animation data, which contain the offset and the frame count
     */
    fun writeImage(name: String, spriteData: List<Tuple<Int>>, paletteOffset: Int, size: Int = 6) {
        val oldImage = SPRITES[name] ?: throw Exception("Null Image for $name")

        // Get the palette from the image
        val palette = (0..15).map { Color(oldImage.getRGB(it, 0)) }
        // Remove the palette so we can write the image
        (1..15).forEach { oldImage.setRGB(it, 0, palette[0].rgb) }

        // Write new palette to rom
        (0..15).forEach {
            val c = GBAColor.toGBA(palette[it])
            val h1 = Integer.parseInt(c.split("(?<=\\G.{2})".toRegex()).dropLastWhile(String::isEmpty)[0], 16)
            val h2 = Integer.parseInt(c.split("(?<=\\G.{2})".toRegex()).dropLastWhile(String::isEmpty)[1], 16)
            SBHS.raf.seek((paletteOffset + it * 2).toLong())
            SBHS.raf.write(h2)
            SBHS.raf.seek((paletteOffset + it * 2 + 1).toLong())
            SBHS.raf.write(h1)
        }

        val img = convertImageToGBAColors(oldImage, palette)

        spriteData.forEachIndexed { animationIndex, animationData ->
            val offset = animationData[0]
            println("0x" + Integer.toHexString(offset))
            val frames = animationData[1]
            var currentFrame = 0
            while (currentFrame < frames) {
                val sections = Array(size * size, { SpriteSection(Array(8, { Array(8, { -1 }) })) })

                var sy = 0
                while (sy < size) {
                    var sx = 0
                    while (sx < size) {
                        var y = 0
                        while (y < 8) {
                            var x = 0
                            while (x < 8) {
                                val ix = sx * 8 + x + (8 * size * animationIndex)
                                val iy = currentFrame * size * 8 + sy * 8 + y
                                sections[sx + sy * size].values[y][x] = try {
                                    val color = Color(img.getRGB(ix, iy))
                                    var value = palette.indexOf(color)
                                    if (value < 0) {
                                        System.err.println("Can't find color $color in the palette.")
                                        value = 0
                                    }
                                    value
                                } catch(e: Exception) {
                                    print("i($ix,$iy)/(${img.width},${img.height});")
                                    println("sx * 8 + x = ${sx * 8 + x}")
                                    println("sy * 8 + y = ${sy * 8 + y}")
                                    println("cf: $currentFrame/$frames for $animationIndex sx: $sx sy: $sy size: $size | ${sx + sy * size} -> ($x, $y)")
                                    e.printStackTrace()
                                    System.exit(-1)
                                    0
                                }
                                x++
                            }
                            y++
                        }
                        sx++
                    }
                    sy++
                }

                /**
                0123OP  012345  00 01 02 03  04 05
                4567QR  6789AB  06 07 08 09  10 11
                89ABST  CDEFGH  12 13 14 15  16 17
                CDEFUV  IJKLMN  18 19 20 21  22 23
                GHIJWX  OPQRST  24 25 26 27  28 29
                KLMNYZ  UVWXYZ  30 31 32 33  34 35

                0, 1, 2, 3,
                6, 7, 8, 9,
                12, 13, 14, 15,
                18, 19, 20, 21,
                24, 25, 26, 27,
                30, 31, 32, 33,
                4, 5, 10, 11,
                16, 17, 22, 23,
                28, 29, 34, 35
                 */
                val sortedSections = arrayOf(
                        sections[0], sections[1], sections[2], sections[3],
                        sections[6], sections[7], sections[8], sections[9],
                        sections[12], sections[13], sections[14], sections[15],
                        sections[18], sections[19], sections[20], sections[21],
                        sections[24], sections[25], sections[26], sections[27],
                        sections[30], sections[31], sections[32], sections[33],
                        sections[4], sections[5], sections[10], sections[11],
                        sections[16], sections[17], sections[22], sections[23],
                        sections[28], sections[29], sections[34], sections[35]
                )

                fun <T> Array<Array<T>>.f(): List<T> {
                    val m = mutableListOf<T>()
                    this.forEach { m.addAll(it) }
                    return m
                }

                val valuesToWrite = mutableListOf<Int>()
                sortedSections.forEachIndexed { s, spriteSection ->
                    val values = spriteSection.values.f().map { Integer.toHexString(it) }
                    val appendedValues = mutableListOf<Int>()

                    var i = 0
                    while (i < values.size) {
                        val v1 = values[i]
                        val v2 = values[i + 1]
                        val v = Integer.parseInt(v2 + v1, 16)
                        appendedValues.add(v)
                        i += 2
                    }

                    valuesToWrite.addAll(appendedValues)
                }

                valuesToWrite.forEachIndexed { i, value ->
                    val o = offset + (currentFrame * size * size * 32) + i.toLong()
                    SBHS.raf.seek(o)
                    SBHS.raf.write(value)
                }

                currentFrame++
            }
        }
    }

    /**
     * Scales image to new size
     * @param src Image to scale
     * *
     * @param w Width of new image
     * *
     * @param h Height of new image
     * *
     * @return Scaled image
     */
    fun scale(src: BufferedImage, w: Int, h: Int): BufferedImage {
        val img = BufferedImage(w, h, BufferedImage.TYPE_INT_RGB)
        var x: Int
        var y: Int
        val ww = src.width
        val hh = src.height
        x = 0
        while (x < w) {
            y = 0
            while (y < h) {
                val col = src.getRGB(x * ww / w, y * hh / h)
                img.setRGB(x, y, col)
                y++
            }
            x++
        }
        return img
    }

    fun BufferedImage.copy(): BufferedImage {
        val cm = this.colorModel
        val raster = this.copyData(this.raster.createCompatibleWritableRaster())
        return BufferedImage(cm, raster, cm.isAlphaPremultiplied, null)
    }

    // TODO: This function is poorly named
    fun BufferedImage.changeColors(l: (Color) -> Color) {
        var y = 0
        while (y < this.height) {
            var x = 0
            while (x < this.width) {
                this.setRGB(x, y, l(Color(this.getRGB(x, y))).rgb)
                x++
            }
            y++
        }
    }

    fun convertImageToGBAColors(img: BufferedImage, palette: List<Color>): BufferedImage {
        // Copy the image so we don't modify the new one
        val newImage = img.copy()
        println("iw${newImage.width}h${newImage.height};")

        // Convert the palette to GBA colors and back
        val newPalette = palette.map { GBAColor.fromGBA(GBAColor.toGBA(it)) }

        // Replace the colors that need to be replaced
        newImage.changeColors {
            if (palette.contains(it)) {
                val index = palette.indexOf(it)
                newPalette[index]
            } else it
        }

        return newImage
    }

    fun printPalette(s: Array<String>) {
        print("[")
        for (i in s.indices) {
            print(s[i] + "(")
            print(GBAColor.fromGBA(s[i]).toString() + ")")
            if (i < s.size - 1) print(", ")
        }
        println("]")
    }

    fun getIndexFromPalette(name: String, color: Color): Int {
        val p = ArrayList(Arrays.asList(*PaletteManager.PALETTES[name]!!))
        for (s in p) {
            val i = p.indexOf(s)
            val pc = GBAColor.fromGBA(s)
            if (pc == color) return i
        }
        print("Error finding color $color ")
        printPalette(PaletteManager.PALETTES[name]!!)
        return 0
    }
}